src/ophyd_async/epics/pmac/_utils.py

```diff
@@ -249,3 +322,139 @@ def calculate_ramp_position_and_duration(
+
+
+def _calculate_profile_from_velocities(
+    motors: list[Motor],
+    time_arrays: dict[Motor, npt.NDArray[np.float64]],
+    velocity_arrays: dict[Motor, npt.NDArray[np.float64]],
+    current_positions: dict[Motor, npt.NDArray[np.float64]],
+) -> tuple[
+    dict[Motor, npt.NDArray[np.float64]],
+    dict[Motor, npt.NDArray[np.float64]],
+    list[float],
+]:
+    """Convert per-axis time/velocity profiles into aligned time/position profiles."""
+    # All unique non‑zero time points across all axes
+    combined_times = np.unique(np.concatenate(list(time_arrays.values())))
+    combined_times = np.sort(combined_times)[1:]  # drop t=0
+    num_intervals = len(combined_times)
+
+    # Convert absolute times to interval durations
+    time_intervals = np.diff(np.concatenate(([0.0], combined_times))).tolist()
+
+    turnaround_profile: dict[Motor, npt.NDArray[np.float64]] = {}
+    turnaround_velocity: dict[Motor, npt.NDArray[np.float64]] = {}
+
+    for motor in motors:
+        axis_times = time_arrays[motor]
+        axis_vels = velocity_arrays[motor]
+        pos = current_positions[motor]
+        prev_vel = axis_vels[0]
+        time_since_axis_point = 0.0
+        axis_pt = 1  # index into this axis's profile
+
+        turnaround_profile[motor] = np.empty(num_intervals, dtype=np.float64)
+        turnaround_velocity[motor] = np.empty(num_intervals, dtype=np.float64)
+
+        for i, dt in enumerate(time_intervals):
+            next_vel = axis_vels[axis_pt]
+            prev_axis_vel = axis_vels[axis_pt - 1]
+            axis_dt = axis_times[axis_pt] - axis_times[axis_pt - 1]
+
+            if np.isclose(combined_times[i], axis_times[axis_pt]):
+                # Exact match with a defined velocity point
+                this_vel = next_vel
+                axis_pt += 1
+                time_since_axis_point = 0.0
+            else:
+                # Interpolate between velocity points
+                time_since_axis_point += dt
+                frac = time_since_axis_point / axis_dt
+                this_vel = prev_axis_vel + frac * (next_vel - prev_axis_vel)
+
+            delta_pos = 0.5 * (prev_vel + this_vel) * dt
+            pos += delta_pos
+            prev_vel = this_vel
+
+            turnaround_profile[motor][i] = pos
+            turnaround_velocity[motor][i] = this_vel
+
+    return turnaround_profile, turnaround_velocity, time_intervals
+
+
+def _get_start_and_end_velocities(
+    motors: list[Motor], slice: Slice, half_durations: npt.NDArray[float64], gap: int
+):
+    start_velocities = {}
+    end_velocities = {}
+    distances = {}
+    for motor in motors:
+        # Velocity from point just before gap (exit velocity)
+        start_velocities[motor] = 2 * (
+            slice.upper[motor][gap - 1] - slice.midpoints[motor][gap - 1]
+        ) / half_durations[gap - 1] - (
+            slice.upper[motor][gap - 1] - slice.lower[motor][gap - 1]
+        ) / (half_durations[gap - 1] * 2)
```

Comment:
```suggestion
        duration1 = half_durations[gap - 1]
        duration2 = half_durations[gap - 1] * 2
        velocity1 = 2 * (slice.upper[motor][gap - 1] - slice.midpoints[motor][gap - 1]) / duration1
        velocity2 = (slice.upper[motor][gap - 1] - slice.lower[motor][gap - 1]) / duration2
        start_velocities[motor] = thing1 - thing2
```

Ideally with descriptive names that explain what we are calculating